### PR TITLE
Add support for `sudo su -` using password auth

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
Allow users to run Ansible tasks through `sudo su -` using password auth

This fixes ansible/ansible#8650 and ansible/ansible#12686

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
sudosu

##### ADDITIONAL INFORMATION
So I have been using this at various customers for bootstrapping Ansible mostly.

Often you have an existing setup where there is a user that has root-access enabled through sudo, but only to run `su` to log using the user's password.
In these specific cases the root password is unique to the system and therefore not an easy way to automate bootstrapping.

Having a `sudo su -` become option **with password prompt** is not possible with the existing become methods (neither sudo nor su can be used) by abusing `become_exe` or `become_flags`.

### CODE REVIEW

File: plugins/become/sudosu.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = """
+    become: sudosu
+    short_description: Run tashks using sudo su -
+    description:
+        - This become plugins allows your remote/login user to execute commands as another user via the sudo and su utility combined.
+    author: ansible (@core)
```
Comment (felixfontein) [inline plugins/become/sudosu.py:12]: The poor person called `core` on GitHub will always get pinged for this :)

You can use `Ansible Core Team` instead. Also you should add yourself to the list.

File: plugins/become/sudosu.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    version_added: "2.12"
```
Comment (felixfontein) [inline plugins/become/sudosu.py:13]:
```suggestion
    version_added: 2.3.0
```

File: plugins/become/sudosu.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    options:
+        become_user:
+            description: User you 'become' to execute the task
```
Comment (felixfontein) [inline plugins/become/sudosu.py:16]:
```suggestion
            description: User you 'become' to execute the task.
```

File: plugins/become/sudosu.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+              - section: sudo_become_plugin
+                key: user
+            vars:
+              - name: ansible_become_user
+              - name: ansible_sudo_user
+            env:
+              - name: ANSIBLE_BECOME_USER
+              - name: ANSIBLE_SUDO_USER
+        become_flags:
+            description: Options to pass to sudo
```
Comment (felixfontein) [inline plugins/become/sudosu.py:30]:
```suggestion
            description: Options to pass to C(sudo).
```

File: plugins/become/sudosu.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+              - section: sudo_become_plugin
+                key: flags
+            vars:
+              - name: ansible_become_flags
+              - name: ansible_sudo_flags
+            env:
+              - name: ANSIBLE_BECOME_FLAGS
+              - name: ANSIBLE_SUDO_FLAGS
+        become_pass:
+            description: Password to pass to sudo
```
Comment (felixfontein) [inline plugins/become/sudosu.py:44]:
```suggestion
            description: Password to pass to C(sudo).
```

File: plugins/become/sudosu.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            required: False
```
Comment (felixfontein) [inline plugins/become/sudosu.py:45]:
```suggestion
            required: false
```

File: plugins/become/sudosu.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2021, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = """
+    become: sudosu
+    short_description: Run tashks using sudo su -
```
Comment (felixfontein) [inline plugins/become/sudosu.py:9]:
```suggestion
    short_description: Run tasks using sudo su -
```

File: plugins/become/sudosu.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    description:
+        - This become plugins allows your remote/login user to execute commands as another user via the sudo and su utility combined.
```
Comment (felixfontein) [inline plugins/become/sudosu.py:11]:
```suggestion
        - This become plugins allows your remote/login user to execute commands as another user via the C(sudo) and C(su) utilities combined.
```

File: plugins/become/sudosu.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                key: password
+"""
+
+
+from ansible.plugins.become import BecomeBase
+
+
+class BecomeModule(BecomeBase):
+
+    name = 'sudo'
```
Comment (felixfontein) [inline plugins/become/sudosu.py:64]:
```suggestion
    name = 'community.general.sudosu'
```

File: plugins/become/sudosu.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    fail = ('Sorry, try again.',)
+    missing = ('Sorry, a password is required to run sudo', 'sudo: a password is required')
+
+    def build_become_command(self, cmd, shell):
+        super(BecomeModule, self).build_become_command(cmd, shell)
+
+        if not cmd:
+            return cmd
+
+        becomecmd = self.name
```
Comment (felixfontein) [inline plugins/become/sudosu.py:76]:
```suggestion
        becomecmd = 'sudo'
```

**Review by felixfontein — COMMENTED** [review]
Thanks for the contribution!

Could you also add some basic unit tests (see https://github.com/ansible-collections/community.general/tree/main/tests/unit/plugins/become/ for examples)?

**Comment by felixfontein** [comment]
CI errors are unrelated to this PR, they come from ansible/ansible#73703.

File: plugins/become/sudosu.py (hunk 1 above)
Comment (dagwieers) [reply plugins/become/sudosu.py:12]: This is in fact a copy of the sudo module. I proposed it for core as IMO it fits better together with sudo and su plugins, but alas...

File: plugins/become/sudosu.py (hunk 2 above)
Comment (dagwieers) [reply plugins/become/sudosu.py:13]: I wouldn't have a clue. Fixed.

File: plugins/become/sudosu.py (hunk 3 above)
Comment (dagwieers) [reply plugins/become/sudosu.py:16]: This is coming from the original sudo plugin, but no issue in adding trailing dots.

File: plugins/become/sudosu.py (hunk 6 above)
Comment (dagwieers) [reply plugins/become/sudosu.py:45]: Would be wrong in sudo as well.

File: plugins/become/sudosu.py (hunk 7 above)
Comment (dagwieers) [reply plugins/become/sudosu.py:9]: Sudo too :)

**Comment by dagwieers** [comment]
I added unit tests (based on sudo) but I have no clue on how to run them from a cloned directory.

File: plugins/become/sudosu.py (hunk 3 above)
Comment (felixfontein) [reply plugins/become/sudosu.py]: There are some plugins/modules in core that someone needs to go through as well, to adjust to all current docs guidelines...

File: tests/unit/plugins/become/test_sudosu.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    sudo_flags = '-H -s -n'
+
+    cmd = play_context.make_become_cmd(cmd=default_cmd, executable=default_exe)
+    assert cmd == default_cmd
+
+    success = 'BECOME-SUCCESS-.+?'
+
+    play_context.become = True
+    play_context.become_user = 'foo'
+    play_context.set_become_plugin(become_loader.get('sudosu'))
```
Comment (felixfontein) [inline tests/unit/plugins/become/test_sudosu.py:34]:
```suggestion
    play_context.set_become_plugin(become_loader.get('community.general.sudosu'))
```

File: tests/unit/plugins/become/test_sudosu.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    play_context.become_flags = sudo_flags
+    cmd = play_context.make_become_cmd(cmd=default_cmd, executable=default_exe)
+
+    assert (re.match("""%s %s  -u %s su -l %s -c 'echo %s; %s'""" % (sudo_exe, sudo_flags, play_context.become_user,
+                                                               default_exe, success, default_cmd), cmd) is not None)
```
Comment (felixfontein) [inline tests/unit/plugins/become/test_sudosu.py:39]:
```suggestion
                                                                     default_exe, success, default_cmd), cmd) is not None)
```

File: tests/unit/plugins/become/test_sudosu.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+    play_context.become_pass = 'testpass'
+    cmd = play_context.make_become_cmd(cmd=default_cmd, executable=default_exe)
+    assert (re.match("""%s %s -p "%s" -u %s su -l %s -c 'echo %s; %s'""" % (sudo_exe, sudo_flags.replace('-n', ''),
+                                                                      r"\[sudo via ansible, key=.+?\] password:", play_context.become_user,
+                                                                      default_exe, success, default_cmd), cmd) is not None)
```
Comment (felixfontein) [inline tests/unit/plugins/become/test_sudosu.py:45]:
```suggestion
                                                                            r"\[sudo via ansible, key=.+?\] password:", play_context.become_user,
                                                                            default_exe, success, default_cmd), cmd) is not None)
```

**Review by felixfontein — COMMENTED** [review]
I think these are the things that prevent the tests from passing:

**Comment by felixfontein** [comment]
> I added unit tests (based on sudo) but I have no clue on how to run them from a cloned directory.

The cloned directory needs to be in the correct directory structure, i.e. it must be called `general`, be inside a directory called `community`, which in turn is inside a directory called `ansible_collections` - so you have `ansible_collections/community/general`. That's basically what Ansible needs to be able to find the collection if it is given the directory containing `ansible_collections` (see also https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html#contributing-to-collections).

**Comment by felixfontein** [comment]
The unit tests still fail, adding some `print` statements when running them locally should show what the problem is... (Galaxy is really annoying today, even worse than usual, so please ignore the other failed runs. I don't want to restart them yet again, since this needs a new commit anyway...)

File: tests/unit/plugins/become/test_sudosu.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    success = 'BECOME-SUCCESS-.+?'
+
+    play_context.become = True
+    play_context.become_user = 'foo'
+    play_context.set_become_plugin(become_loader.get('community.general.sudosu'))
+    play_context.become_flags = sudo_flags
+    cmd = play_context.make_become_cmd(cmd=default_cmd, executable=default_exe)
+
+    assert (re.match("""%s %s  -u %s su -l %s -c 'echo %s; %s'""" % (sudo_exe, sudo_flags, play_context.become_user,
+                                                                     default_exe, success, default_cmd), cmd) is not None)
```
Comment (felixfontein) [inline tests/unit/plugins/become/test_sudosu.py:39]:
```suggestion
    assert (re.match("""%s %s  su -l %s %s -c 'echo %s; %s'""" % (sudo_exe, sudo_flags, play_context.become_user,
                                                                  default_exe, success, default_cmd), cmd) is not None)
```

File: tests/unit/plugins/become/test_sudosu.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+    play_context.become_pass = 'testpass'
+    cmd = play_context.make_become_cmd(cmd=default_cmd, executable=default_exe)
+    assert (re.match("""%s %s -p "%s" -u %s su -l %s -c 'echo %s; %s'""" % (sudo_exe, sudo_flags.replace('-n', ''),
+                                                                            r"\[sudo via ansible, key=.+?\] password:", play_context.become_user,
+                                                                            default_exe, success, default_cmd), cmd) is not None)
```
Comment (felixfontein) [inline tests/unit/plugins/become/test_sudosu.py:45]:
```suggestion
    assert (re.match("""%s %s -p "%s" su -l %s %s -c 'echo %s; %s'""" % (sudo_exe, sudo_flags.replace('-n', ''),
                                                                         r"\[sudo via ansible, key=.+?\] password:", play_context.become_user,
                                                                         default_exe, success, default_cmd), cmd) is not None)
```

**Review by felixfontein — COMMENTED** [review]
With these changes, the unit tests pass. @dagwieers could you please check whether this is what you intended the plugin to do?

File: plugins/become/sudosu.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+__metaclass__ = type
+
+DOCUMENTATION = """
+    become: sudosu
+    short_description: Run tasks using sudo su -
+    description:
+        - This become plugins allows your remote/login user to execute commands as another user via the C(sudo) and C(su) utilities combined.
+    author:
+    - Dag Wieers (@dagwieers)
+    version_added: 2.3.0
```
Comment (felixfontein) [inline plugins/become/sudosu.py:14]:
```suggestion
    version_added: 2.4.0
```

**Comment by dagwieers** [comment]
> > I added unit tests (based on sudo) but I have no clue on how to run them from a cloned directory.
> 
> The cloned directory needs to be in the correct directory structure, i.e. it must be called `general`, be inside a directory called `community`, which in turn is inside a directory called `ansible_collections` - so you have `ansible_collections/community/general`. That's basically what Ansible needs to be able to find the collection if it is given the directory containing `ansible_collections` (see also https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html#contributing-to-collections).

That is a major pain. You ought to be able to test a collection from a cloned directory IMO.

**Comment by dagwieers** [comment]
> With these changes, the unit tests pass. @dagwieers could you please check whether this is what you intended the plugin to do?

Thanks for this. I made the wrong edits, expecting to be able to test it :-)
For me this looks fine, I tested it once more and all is well.

**Comment by felixfontein** [comment]
> That is a major pain. You ought to be able to test a collection from a cloned directory IMO.

It's the most common problem with running tests I think. I guess that ansible-test could be adjusted to work without that directory structure if `--docker` is used, but without `--docker` this is probably next to impossible without making modifications to ansible-core itself. (Except if copying all needed files to /tmp into the correct directory structure would be acceptable.)

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@dagwieers thanks a lot for contributing this!

**Pull request merged**